src/librustc_typeck/check/mod.rs

```diff
@@ -3751,8 +3751,40 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                  error_code: &str,
                                  c_variadic: bool,
                                  sugg_unit: bool| {
+            let (span, arg_spans) = match &expr.kind {
+                hir::ExprKind::Call(hir::Expr { span, .. }, args) => (
+                    *span,
+                    if args.is_empty() {
+                        // foo()
+                        // ^^^-- supplied no parameters
+                        // |
+                        // expected 2 parameters
+                        vec![tcx.sess.source_map().next_point(*span).with_hi(sp.hi())]
+                    } else {
+                        // foo(1, 2, 3)
+                        // ^^^ -  -  - supplied 3 parameters
+                        // |
+                        // expected 2 parameters
+                        args.iter().map(|arg| arg.span).collect::<Vec<Span>>()
+                    },
+                ),
+                hir::ExprKind::MethodCall(_, span, args) => (
+                    *span,
+                    if args.is_empty() {
+                        // `sp` doesn't point at the whole `foo.bar()`, only at `bar`.
```

Comment:
Weird. Perhaps it *should* point to the whole `foo.bar()`?
Then we'll be able to use the same `vec![tcx.sess.source_map().next_point(*span).with_hi(sp.hi())]` as `ExprKind::Call`.